src/dag/key.rs

```diff
@@ -4,43 +4,47 @@ use std::str;
 // KVKey is the key we use to store our dag data in the underlying
 // kvstore.
 #[derive(Debug, PartialEq, Eq)]
-pub enum Key {
-    ChunkData(String),
-    ChunkRefs(String),
-    Head(String),
+pub enum Key<'a> {
```

Comment:
Nice work!

Reply:
I did read the life time section and just wanted to make sure I understand this change.

`String` is heap allocated. `str` is stack allocated. `'a` parameter means that the lifetime of `Key` must be shorter than the arguments to `ChunkData`/`ChunkMeta`/`Head`?

Reply:
It's not a question of where it is allocated, it's a question of ownership. `String` owns its internal storage. Think of it like `scoped_ptr<std::string>`. `str` is a slice -- it's just a reference to some other storage elsewhere. Think of it like `StringPiece` from Chrome. That storage could be heap allocated or stack.

The lifetime parameters here say:

* There is a struct `Key`
* Which is parameterized by the abstract lifetime `'a` (looking at just this code we don't know what that lifetime is, just that it exists)
* Each of the three fields are a reference to a string elsewhere
* Those three fields all have the same lifetime, the lifetime that is passed in when this type is instanciated

In practice, this is used to avoid having to copy the string data into this struct when it is created. Instead, we can just have its fields reference string data that lives elsewhere.

Reply:
(i think -- @whiten want to crosscheck my explanation?? :))

Reply:
(updated -- this is an enum not a struct)

Reply:
Ha, thanks for going first here. I think that's all right, and I just want to emphasize your statement that they "all have the same lifetime", there's no inequality here. But, "they all" is referring to the Key struct and the *references* to the str, not the underlying u8 Vec that backs it.

Thus, anyone with access to the Key could take another reference to that data, and the compiler must assert that it's dropped before the Key is. And, in the Key::parse() function, the compiler can allow a Key with references to the &str that was passed in to escape because the Key is defined to last no longer than those references.

It seems that when we name lifetimes we express lifetime equalities, but with the purpose of allowing the compiler to prove lifetime inequalities, e.g. that the underlying str lasts as long as or longer than the Key that references it.